### PR TITLE
proposal: update `Support restoring SQL text from an AST tree` proposal for new interface

### PR DESCRIPTION
**Pull request by leoppro**
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8694)

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -3,7 +3,7 @@ language: go
 go_import_path: github.com/pingcap/tidb
 
 go:
-  - "1.11.2"
+  - "1.11.3"
```
Comment (zz-jason) [inline .travis.yml]: irrelevant change?

Comment (leoppro) [reply .travis.yml]: oh, this afternoon our Travis CI have a error.

Comment (zz-jason) [reply .travis.yml]: so we have to update `go` to fix that?

**Comment by XuHuaiyu** [comment]
PTAL @CaitinChen

File: docs/design/2018-11-29-ast-to-sql-text.md [hunk 2]
```diff
@@ -30,14 +30,33 @@ I recommend adding a `Restore()` method to ast.Node. This is the function define
...
+
+[RestoreStringSingleQuotes, RestoreStringDoubleQuotes]  
+[RestoreStringEscapeBackslash]  
+[RestoreKeyWordUppercase, RestoreKeyWordLowercase]  
+[RestoreNameUppercase, RestoreNameLowercase]  
+[RestoreNameDoubleQuotes, RestoreNameBackQuotes]  
+The flag with the left position in each group has a higher priority.  
+
+This is `RestoreCtx` struct: 
+
```
Comment (XuHuaiyu) [inline docs/design/2018-11-29-ast-to-sql-text.md]: Please add the default RestoreFlags definition.

File: .travis.yml (hunk 1 above)
Comment (leoppro) [reply .travis.yml]: The conclusion is that the go 1.11.4 can't compile tidb successfully

File: docs/design/2018-11-29-ast-to-sql-text.md [hunk 3]
```diff
@@ -30,14 +30,35 @@ I recommend adding a `Restore()` method to ast.Node. This is the function define
...
+	// Restore AST to SQL text and append them to `ctx`.
 	// return error when the AST is invalid.
-	Restore(sb *strings.Builder) error
+	Restore(ctx *RestoreCtx) error
 	
 	...
 }
 ```
 
+In order to output sql text in different formats, I introduced nine flags. There are mutually exclusive group of `RestoreFlags`:
```
[the remark this answers is not in the document]
Comment (leoppro) [reply docs/design/2018-11-29-ast-to-sql-text.md]:
```

[the remark this answers is not in the document]
Comment (CaitinChen) [reply docs/design/2018-11-29-ast-to-sql-text.md]: @leoppro Yes

**Comment by CaitinChen** [comment]
LGTM

**Pull request merged**